glue/src/main/scala/com/gu/tableversions/glue/GlueMetastore.scala

```diff
@@ -0,0 +1,184 @@
+import com.gu.tableversions.metastore.Metastore.TableOperation.{
+  AddPartition,
+  RemovePartition,
+  UpdatePartitionVersion,
+  UpdateTableVersion
+}
+import com.gu.tableversions.metastore.{Metastore, VersionPaths}
+
+import scala.collection.JavaConversions._
+
+class GlueMetastore[F[_]](glue: AWSGlue)(implicit F: Sync[F]) extends Metastore[F] {
+
+  override def currentVersion(table: TableName): F[TableVersion] = {
+
+    def getPartitionColumns(glueTable: GlueTable): List[PartitionColumn] =
+      Option(glueTable.getPartitionKeys).toList.flatten map { glueColumn =>
+        PartitionColumn(glueColumn.getName)
+      }
+
+    def snapshotTableVersion(tableLocation: URI): TableVersion =
+      SnapshotTableVersion(VersionPaths.parseVersion(tableLocation))
+
+    for {
+      glueTable <- getGlueTable(table)
+      partitionColumns <- F.delay(getPartitionColumns(glueTable))
+      tableLocation <- F.delay(findTableLocation(glueTable))
+      version <- if (partitionColumns.isEmpty)
+        F.delay(snapshotTableVersion(tableLocation))
+      else
+        getPartitionedTableVersion(table, tableLocation, partitionColumns)
+    } yield (version)
+  }
+
+  private def getPartitionedTableVersion(
+      table: TableName,
+      tableLocation: URI,
+      partitionColumns: List[PartitionColumn]): F[TableVersion] = {
+
+    def toPartition(columnValues: List[ColumnValue]): Partition = columnValues match {
+      case head :: tail => Partition(head, tail: _*)
+      case _            => throw new Exception("empty columnValues list for partition found")
+    }
+
+    def toPartitionWithVersion(gluePartition: GluePartition): (Partition, Version) = {
+
+      val partitionColumnAndValue: List[(PartitionColumn, String)] = partitionColumns.zip(gluePartition.getValues)
+      val columnValues: List[ColumnValue] = partitionColumnAndValue.map { case (key, value) => ColumnValue(key, value) }
+      val partition: Partition = toPartition(columnValues)
+      val location = new URI(gluePartition.getStorageDescriptor.getLocation)
+      partition -> VersionPaths.parseVersion(location)
+    }
+
+    getGluePartitions(table).map { gluePartitions =>
+      val partitionVersions: Map[Partition, Version] = gluePartitions.map(toPartitionWithVersion).toMap
+      PartitionedTableVersion(partitionVersions)
+    }
+
+  }
+
+  override def update(table: TableName, changes: Metastore.TableChanges): F[Unit] =
+    changes.operations.traverse_(appliedOp(table))
+
+  private def appliedOp(table: TableName)(operation: TableOperation): F[Unit] =
+    operation match {
+      case AddPartition(partition, version)           => addPartition(table, partition, version)
+      case UpdatePartitionVersion(partition, version) => updatePartitionVersion(table, partition, version)
+      case RemovePartition(partition)                 => removePartition(table, partition)
+      case UpdateTableVersion(versionNumber)          => updateTableLocation(table, versionNumber)
+    }
```

Comment:
It's interesting that the two methods above are the same for this implementation and the Spark/Hive one. We could consider pulling this out into the base trait in a later PR, similar to what I did for `TableVersions`.

Reply:
yeah that sounds reasonable.